java/algorithm/src/main/java/org/apache/arrow/algorithm/search/VectorSearcher.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.algorithm.search;
+
+import org.apache.arrow.algorithm.sort.VectorValueComparator;
+import org.apache.arrow.vector.ValueVector;
+
+/**
+ * Search for a particular element in the vector.
+ * @param <V> vector type.
+ */
+public abstract class VectorSearcher<V extends ValueVector> {
```

Comment:
an abstract class might be overkill in this case.  Did you consider adding a utility class with static methods (1 for unsorted and one for sorted)?  Do you expect many more implementations?  It might also be surprising to consumers that use this in there API that it doesn't work properly (and fails silenty) in some cases if the data isn't sorted.

Reply:
For this particular PR, an abstract super class is truly a overkill, and it is a good idea to implement them as static methods. 

I think more searcher implementations will be added in the future, including, for example, bucket based searcher, search tree based searcher, hash table based searcher, B-tree based searcher, etc.

For me, either way is OK (add an abstract class or implement as static methods). Maybe we can add the abstract class later in the future, when we implement more search algorithms. So please give some comments.

For the problem of unexpected search results in sorted searcher, do you think we should have a method to (optionally) check the vector order in O(n) time?

Reply:
I think for now I would prefer static methods.

A method to check if a vector is sorted potentially useful, but I don't think it should be incorporated here (it can be another PR).

Reply:
Sure. Thanks for your feedback. 
I have revised the code to use static methods.